### PR TITLE
chain sync modes

### PR DESCRIPTION
**Pull request by frrist**
This PR is a sketch of adding a Bootstrap and Caught-Up sync mode to the syncer. There are no plans to merge this, just here to express ideas and show areas that may need work/refactoring.

### CODE REVIEW

File: chain/peer_manager.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func (bpm *BasicPeerManager) SelectHead() (types.TipSet, error) {
+	bpm.peerHeadMu.Lock()
+	defer bpm.peerHeadMu.Unlock()
+
+	if len(bpm.peerHeads) == 0 {
+		return types.UndefTipSet, fmt.Errorf("zero peers to select from")
+	}
+	// for now pick something random?
```
Comment (frrist) [inline chain/peer_manager.go]: Should probably look through the list of heads and return the heaviest.

File: node/node.go [hunk 2]
```diff
@@ -428,7 +429,7 @@ func (nc *Config) Build(ctx context.Context) (*Node, error) {
 	fcWallet := wallet.New(backend)
 
 	// only the syncer gets the storage which is online connected
-	chainSyncer := chain.NewSyncer(&cstOffline, nodeConsensus, chainStore, fetcher, chain.Syncing)
+	chainSyncer := chain.NewSyncer(&cstOffline, nodeConsensus, chainStore, fetcher, chain.Bootstrap)
```
Comment (frrist) [inline node/node.go]: When we create a syner for the node we place it in Bootstrap mode

File: node/block.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ func (node *Node) processBlock(ctx context.Context, pubSubMsg pubsub.Message) (e
 	// Don't be too quick to change that, though: the syncer re-fetching the block
 	// is currently critical to reliable validation.
 	// See https://github.com/filecoin-project/go-filecoin/issues/2962
-	err = node.Syncer.HandleNewTipset(ctx, types.NewTipSetKey(blk.Cid()))
+	err = node.Syncer.HandleNewTipset(ctx, pubSubMsg.GetFrom(), types.NewTipSetKey(blk.Cid()))
```
Comment (frrist) [inline node/block.go]: We include who the block is from so we can populate the PeerManager used by the syncer

File: consensus/block_validation.go [hunk 4]
```diff
@@ -71,6 +71,10 @@ func (dv *DefaultBlockValidator) ValidateSemantic(ctx context.Context, child *ty
 
 // ValidateSyntax validates a single block is correctly formed.
 func (dv *DefaultBlockValidator) ValidateSyntax(ctx context.Context, blk *types.Block) error {
+	// hack validate genesisblock
```
Comment (frrist) [inline consensus/block_validation.go]: Unrelated to these changes -- we will probably want some special handling for genesis block validation.

File: chain/syncer_internal_test.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
+package chain
```
Comment (frrist) [inline chain/syncer_internal_test.go]: ignore this entire file

File: chain/syncer.go [hunk 6]
```diff
@@ -419,24 +433,166 @@ func (syncer *Syncer) widen(ctx context.Context, ts types.TipSet) (types.TipSet,
...
+	if !(len(syncer.pm.Peers()) >= BootstrapThreshold) {
+		logSyncer.Infof("Unable Bootstrap Sync, not enough peers to bootstrap from, current count %d", len(syncer.pm.Peers()))
+		return errors.New("Not enough peers to bootstrap")
+	}
+
+	go func() {
+		syncer.mu.Lock()
+		defer syncer.mu.Unlock()
+
+		switch syncer.syncMode {
```
Comment (frrist) [inline chain/syncer.go]: Where we make the decision on how to sync

**Comment by anorth** [comment]
@frrist could you please rebase on master? I pushed over my branch so the diff is messed up now

**Comment by frrist** [comment]
A lot has changed under the branch (it was built on top of some changes that have recently landed) closing in favor of: https://github.com/filecoin-project/go-filecoin/pull/3081

**Pull request closed without merging**